numpy/core/src/umath/ufunc_object.c

```diff
@@ -5893,13 +5893,55 @@ new_array_op(PyArrayObject *op_array, char *data)
     return (PyArrayObject *)r;
 }
 
+/*
+ * Use an indexed loop to do the work
+ * Returns 0 if successful
+ */
+static int
+trivial_at_loop(PyArrayMethodObject *ufuncimpl, NPY_ARRAYMETHOD_FLAGS flags,
+                    PyArrayMapIterObject *iter,
+                    PyArrayObject *op1_array, PyArrayObject *op2_array,
+                    PyArrayMethod_Context *context)
+{
+    int buffersize=0, errormask = 0;
+    int res;
+    char *args[3];
+    npy_intp dimensions = iter->size;
+    npy_intp steps[3];
+    args[0] = (char *) iter->baseoffset;
+    args[1] = (char *) iter->outer_ptrs[0];
+    args[2] = (char *)PyArray_DATA(op2_array);
+    steps[0] = iter->fancy_strides[0];
+    steps[1] = iter->outer_strides[0];
+    steps[2] = PyArray_STRIDE(op2_array, 0);
+    res = ufuncimpl->contiguous_indexed_loop(context, args, &dimensions, steps, NULL);
```

Comment:
Hmmm... Because currently I think we only ever take this when there are two operands, the fact that `op2_array` is passed implicitly asserts that the index array is also 1-D.

Here, we use that fact:  If it was not, I am not certain that the index array couldn't a 2-D and non-contiguous.
It would be nice to at least assert that fact here, because if you add single operand loops, I am not sure that things may not fail in rather strange ways.

Reply:
@seberg - this looks nice, and definitely is something I had completely missed. I think it would be good to add a comment (perhaps on top) what conditions the function expects. It now assumes that operator 0 cannot change (which I think is correct, but good to be explicit about why that is -- cannot need casting since the operation is in-place; and operator 2 needing casting would give the slow path).